### PR TITLE
NeuralNetDiffEq GSoC Blog

### PR DESCRIPTION
**Pull request by akaysh**

### CODE REVIEW

**Comment by akaysh** [comment]
@ChrisRackauckas @oxinabox

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
+---
+layout: post
+title: "NeuralNetDiffEq.jl: A Neural Network solver for Ordinary Differential Equations."
+author: Akshay Sharma
+---
+
+My GSoC 2007 [project](https://summerofcode.withgoogle.com/projects/#5850956641075200) was to implement a package for Julia to solve Ordinary Differential Equations using Neural Networks. The purpose of the project was to provide an additional DE solver using Neural Networks which have a few parallelism and generalization advantages over other solvers. The project was based on a [research paper](https://arxiv.org/pdf/physics/9705023.pdf) which proposed the function approximation capabilities of Neural Networks for solving differential equations. The project was a mixture of research as well as implementation aspects and still has a few parts left to work upon. I chose to work on this project as I have interest in Mathematics and Machine Learning and it involved concepts of both the fields. The package uses [DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) for the solver interface and [KNet.jl](https://github.com/denizyuret/Knet.jl) for NN solver implementation.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "The purpose of the project was to provide an additional DE solver using Neural Networks which have a few parallelism and generalization advantages over other solvers.". It's no more general than the other solvers, but the parallelism in time is the key advantage.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+## Research aspect of the project and The Challenge
+The research paper we referred on the topic is quite old and lacks proper explanation as well as examples. The research done on this part of Neural Networks is not too much and we were not able to get much help from the research papers related to the topic. The initial task was to read and understand the mathematics behind solving Differential Equations. Also the (analytical) methods used to solve Differential Equations on computers are quite different from the ones we use on paper so it took quite some time to get familiar with them. Also the structure and type of NN to be used so that the solver advantages are retained without compromising the performance was a research subdomain as well as a challenge.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Watch out for "also also".

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+Also after implementing the solver for ODEs (Ordinary Differential Equations) and System of ODEs the difficult part was to make the Neural Network converge, especially for the system of ODEs. As there are a lot of factors involved in Neural Networks like Hidden Layer Width, Number of Hidden Neurons, Activations, Weights etc., I relied on my Machine Learning background as well as the help from my mentors to experiment with most of the feasible settings of NN hyper-parameters and recording the accuracy of convergence and performance of the solver.
+
+Making NN converge for System of ODEs is not as easy as it seems and took up most of the time for experimentation and tuning. Predicting the system of DEs solution with larger domain is still a challenge which needs to be worked upon.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "Predicting the system of DEs solution with larger domain is still a challenge which needs to be worked upon.". I don't understand what you mean here.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+## Implementation and Work
+Implementation involved integration of Mathematical and Machine Learning aspects through programming and to build a solver for ODEs. The [DiffEqBase library](https://github.com/JuliaDiffEq/DiffEqBase.jl) is used as a base to extend the Algorithm and solver interface. The work done till now can be seen on the [github repository](https://github.com/JuliaDiffEq/NeuralNetDiffEq.jl) primarily in [this branch](https://github.com/JuliaDiffEq/NeuralNetDiffEq.jl/tree/SingleNN_Approach). This work involves implementing a Neural Network solver for ODEs with customized iterpolation based on NN prediction. The solver also works for system of ODEs but needs to be optimized for larger domains.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "but needs to be optimized for larger domains." I don't know what you mean here.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+### How does it work?
+We need to construct a trial solution for our Differential Equation in terms of the NN output which should also satisfy the DE boundary conditions.  
+We define a loss function for the Neural Network which is same as equating the Differential to zero and minimizing that function (closer to 0 better convergence) wuth the trial solution substituted in it in place of the original function (or the solution to the DE). The Neural Network uses Adam optimization algorithm for backpropagation to tune its weights. GPU operation is supported by using the KnetArray type (the arry type used in [KNet.jl](https://github.com/denizyuret/Knet.jl)) instead of regular Array for parameters and data, but requires CUDNN driver installed to access GPU hardware.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: To do so, we define ...

Comment (ChrisRackauckas) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: equating the derivative to zero

Comment (ChrisRackauckas) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: The opening sentence in general is a little confusing. I would be more direct: the loss function is the difference between the derivative of the neural net solution and the true derivative defined by the ODE.

Comment (ChrisRackauckas) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "GPU operation is supported" odd construction.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+
+
+    lotka_volterra (generic function with 1 method)
+
+
+
+
+```julia
+prob_ode_lotkavoltera = ODEProblem(lotka_volterra,Float32[1.0,1.0],(Float32(0.0),Float32(1.0)))
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: You should try to solve this on `(0.0,10.0)`. This is a classic oscillating equation and it looks like it's incorrect because the solution isn't oscillating, when in reality its timespan is just too short.

Comment (ChrisRackauckas) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: (0.0,5.0) is probably enough if it needs to be faster.

**Comment by ChrisRackauckas** [comment]
You should break up the paragraphs so that way you don't have such long line lengths. I'm not talking about separating them in the rendered result, just the "behind the scenes" `.md` text. (Did I make sense?)

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md (hunk 3 above)
Comment (akaysh) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: I was referring to the ode problem domain which we solve for
Example:
`prob_ode_lotkavoltera = ODEProblem(lotka_volterra,Float32[1.0,1.0],(Float32(0.0),Float32(1.0)))`
and 
`prob_ode_lotkavoltera = ODEProblem(lotka_volterra,Float32[1.0,1.0],(Float32(0.0),Float32(10.0)))
`

Comment (oxinabox) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: You might want to fully highlight that with that example.
Eg showing "While lotka-voltera works over the short domain of 0-1 (Plot), we can see it fails over a larger domain e.g. 0-10 (Plot)"

Nothing is wrong with highlighting the failure of the method clearly.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md (hunk 6 above)
Comment (akaysh) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Do not have an optimize NN for that but will update later when I have.
The graph won't be proper now

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md (hunk 3 above)
Comment (akaysh) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Alright!

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
+---
+layout: post
+title: "NeuralNetDiffEq.jl: A Neural Network solver for Ordinary Differential Equations."
+author: Akshay Sharma
+---
+
+My GSoC 2007 [project](https://summerofcode.withgoogle.com/projects/#5850956641075200) was to implement a package for Julia to solve Ordinary Differential Equations using Neural Networks. The purpose of the project was to provide an additional DE solver using Neural Networks which has parallelism in time as the key advantage over other solvers which are iterative in nature. The project was based on a [research paper](https://arxiv.org/pdf/physics/9705023.pdf) which proposed the function approximation capabilities of Neural Networks for solving differential equations. The project was a mixture of research as well as implementation aspects and still has a few parts left to work upon. I chose to work on this project as I have interest in Mathematics and Machine Learning and it involved concepts of both the fields. The package uses [DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) for the solver interface and [KNet.jl](https://github.com/denizyuret/Knet.jl) for NN solver implementation.
```
Comment (rohitvarkey) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: /s/2007/2017

Comment (akaysh) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Any more changes required before merging?

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md (hunk 4 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Yes, this should be addressed and then I think it's ready.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  ```
+
+
+
+
+      1-element Array{Any,1}:
+        Any[0.574705]
+
+
+### SODE Examples
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Don't call this an SODE. It's not an SODE. It's just a system of ODEs.

**Comment by ChrisRackauckas** [comment]
I think fixing line lengths is the final touch. Any comments @oxinabox ?

**Comment by akaysh** [comment]
What do you mean by fixing lengths? @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
Like when coding, don't make it one big block of text. Instead put new lines in. In Markdown a paragraph takes 2 new lines, so this won't effect the rendering but it will make the backend code nicer. It's a small detail but since everything else tends to follow it, you should too.

**Comment by oxinabox** [comment]
I'll take a final careful read today.

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
+---
+layout: post
+title: "NeuralNetDiffEq.jl: A Neural Network solver for Ordinary Differential Equations."
+author: Akshay Sharma
+---
+
+My GSoC 2007 [project](https://summerofcode.withgoogle.com/projects/#5850956641075200) was to implement a package for Julia to solve Ordinary Differential Equations using Neural Networks.
+The purpose of the project was to provide an additional DE solver using Neural Networks which has parallelism in time as the key advantage over other solvers which are iterative in nature. The project was based on a [research paper](https://arxiv.org/pdf/physics/9705023.pdf) which proposed the function approximation capabilities of Neural Networks for solving differential equations. The project was a mixture of research as well as implementation aspects and still has a few parts left to work upon.
```
Comment (oxinabox) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "The project was based on the research paper of [<author> <year>]"

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+I chose to work on this project as I have interest in Mathematics and Machine Learning and it involved concepts of both the fields. The package uses [DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) for the solver interface and [KNet.jl](https://github.com/denizyuret/Knet.jl) for NN solver implementation.
+
+## How to use Neural Network for solving Differential Equations?
+The concept of this solver is based on the UAT (Universal Approximation Theorem) which says that a Neural Network with not more than 3 hidden can approximate any continuous function. The Neural Network is made to minimize a loss function which then results in the convergence of our Trial solution towards the actual (analytical) solution. To know more about UAT [click here](http://neuralnetworksanddeeplearning.com/chap4.html).
```
Comment (oxinabox) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "which says a neural network with at least 1 hidden layer can..."

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+## Research aspect of the project and The Challenge
+The research paper we referred on the topic is quite old and lacks proper explanation as well as examples. The research done on this part of Neural Networks is not too much and we were not able to get much help from the research papers related to the topic.
```
Comment (oxinabox) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: Better might be to say that understanding the examples and explanations from the paper was challenging.
It actually had a fair few examples, but they were not easy to handle

File: blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+The initial task was to read and understand the mathematics behind solving Differential Equations. Also the (analytical) methods used to solve Differential Equations on computers are quite different from the ones we use on paper so it took quite some time to get familiar with them. The structure and type of NN to be used so that the solver advantages are retained without compromising the performance was a research subdomain as well as a challenge.
```
Comment (oxinabox) [inline blog/_posts/2017-09-04-gsoc-NeuralNetDiffEq.md]: "Also the computational methods used to solve Differential Equations are quite different from the ones we use on paper"

**Comment by ChrisRackauckas** [comment]
Only blocking parts are the requested changes from @oxinabox

**Comment by ViralBShah** [comment]
Assuming it is good to go, it will need a date update to be current before merging.

**Comment by ChrisRackauckas** [comment]
Fixed the date. I'll take one more read and merge.

**Comment by ChrisRackauckas** [comment]
`.DS_Store` these need to be deleted.

**Comment by ChrisRackauckas** [comment]
@oxinabox I did some extensive edits, could you give it a quick read and then we'll merge?

File: blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+This work involves implementing a Neural Network solver for ODEs with customized interpolation based on NN prediction. 
+
+### How does it work?
+We construct a trial solution for our differential equation in terms of the NN output which should also satisfy the DE boundary conditions.
+We define a loss function for the neural net which is the difference between the derivative of the neural net solution and the true derivative defined by the ODE.
+This loss function is minimized (by equating the derivative difference to zero) using the NN (closer to 0 better convergence) with the trial solution substituted in it in place of the original function (or the solution to the DE). The Neural Network uses Adam optimization algorithm for backpropagation to tune its weights.
```
Comment (oxinabox) [inline blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md]: The Neural Network tunes its weights using the Adam optimization algorithm on the backpropagated gradients from that loss function.

File: blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+## Implementation and Work
+The implementation involved integration of mathematical and machine learning aspects to build a neural net solver for ODEs. The [DiffEqBase library](https://github.com/JuliaDiffEq/DiffEqBase.jl) is used as a base to extend the algorithm and solver interface while the neural network was developed using the [Knet.jl library](https://github.com/denizyuret/Knet.jl). The work done till now can be seen on the [NeuralNetDiffEq.jl github repository](https://github.com/JuliaDiffEq/NeuralNetDiffEq.jl), primarily in [this branch](https://github.com/JuliaDiffEq/NeuralNetDiffEq.jl/tree/SingleNN_Approach).
+This work involves implementing a Neural Network solver for ODEs with customized interpolation based on NN prediction. 
+
+### How does it work?
+We construct a trial solution for our differential equation in terms of the NN output which should also satisfy the DE boundary conditions.
+We define a loss function for the neural net which is the difference between the derivative of the neural net solution and the true derivative defined by the ODE.
```
Comment (oxinabox) [inline blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md]: make that "the derivative of the neural net solution with regards to its input"

Add after:

"This is an unusual loss function, in that in includes the gradient of the network itself. It is almost unseen elsewhere in other ML applications"

File: blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md (hunk 13 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-10-12-gsoc-NeuralNetDiffEq.md]: much better

**Comment by ChrisRackauckas** [comment]
I updated the date and fixed it up to match @oxinabox's last 2 comments. Looks like that means it's good to merge @ViralBShah ?
